### PR TITLE
Changed the Nanopb Macro To A Rule

### LINKED ISSUE #1133: Rewrite NanoPb Bazel Macro To Be A Bazel Rule
### Description of the task

We need to manually compile nanopb for Bazel. Currently the Macro does not have Bazel context so we cannot expand on targets. (i.e we have to do

~~~skylark
# c library (TODO since this a macro we lack context to use :fw_proto) check w/ Gareth
c_proto_library(
    name = "fw_c_proto",
    srcs = proto_files,
)
~~~

When we would like to do something like this:

~~~skylark
# c library (TODO since this a macro we lack context to use :fw_proto) check w/ Gareth
c_proto_library(
    name = "fw_c_proto",
    deps = [":fw_proto"],
)
~~~

Where the `fw_proto` is a proto library. 

More info here: https://docs.bazel.build/versions/master/skylark/lib/ctx.html

### Acceptance criteria

- [ ] Change macro to be a proper bazel rule so it can take `deps` instead of `srcs`
- [ ] Convert `file_group`  in `firmware_new/proto/BUILD` to a `proto_library`
- [ ] Investigate and change use of relative vs abs path in macro `nanopb.bzl`
- [ ] Be able to build protobuf defined over multiple files w/ include statements

### Blocked By
Nothing

### PR DESCRIPTION
**Pull request by garethellis0**
### Resolved Issues
resolves #1133 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/firmware_new/proto/control.proto [hunk 1]
```diff
@@ -1,5 +1,7 @@
 syntax = "proto2";
 
+//import "software/proto/grSim_commands.proto";
```
Comment (MathewMacDougall) [inline src/firmware_new/proto/control.proto]: why is this here?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

**Review by jonathanlew — CHANGES_REQUESTED** [review]
CI is failing btw

File: src/firmware_new/nanopb.bzl [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    (compilation_context, compilation_outputs) = cc_common.compile(
+        name = "compile_nanopb_outputs",
+        actions = ctx.actions,
+        feature_configuration = feature_configuration,
+        cc_toolchain = cc_toolchain,
+        srcs = all_proto_src_files,
+        public_hdrs = all_proto_hdr_files,
+        # TODO: this really feels like a bit of a hack. Is there no cleaner way to
+        #       construct this path
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/firmware_new/nanopb.bzl]: Removed, I could not find a better way to do this.

File: src/firmware_new/proto/control.proto (hunk 1 above)
Comment (garethellis0) [reply src/firmware_new/proto/control.proto]: Whoops, there should be no changes to this file. Removed.

File: src/firmware_new/proto/BUILD [hunk 3]
```diff
@@ -1,30 +1,17 @@
...
-# add more proto here
-proto_files = [
-    "control.proto",
-]
-
-# proto files
 proto_library(
     name = "fw_proto",
-    srcs = proto_files,
+    srcs = ["control.proto"],
```
Comment (akhilveeraghanta) [inline src/firmware_new/proto/BUILD:8]: could you try building `tbots_robot_msg.proto` and `components.proto`? Thats the big one because of the `#include` statements, couldn't get that to work easily with the macro

File: src/firmware_new/nanopb.bzl [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            inputs = all_proto_files,
+            outputs = [pb_file],
+            arguments = ["-o", pb_file.path, proto_file.path],
+            executable = ctx.executable.protoc,
+            mnemonic = "ProtoCompile",
+            use_default_shell_env = True,
+        )
+
+        # Generate the equivalent C code using Nanopb
+        h_out_name = generation_folder_name + proto_file.path[:-len(".proto")] + ".pb.h"
```
Comment (akhilveeraghanta) [inline src/firmware_new/nanopb.bzl:34]: `udp_multicast` is currently expecting `_fw.pb.h` for nanopb generated headers, I think you are going to need to append something here, because both C and C++ generate `.pb.h` files

that also seems to be the reason for the CI failure

**Review by akhilveeraghanta — COMMENTED** [review]
nice work 🔥

File: src/firmware_new/proto/BUILD (hunk 3 above)
Comment (garethellis0) [reply src/firmware_new/proto/BUILD:8]: Sorry, should have mentioned this above, but I did confirm that `#include`'s work with this rule.

File: src/firmware_new/nanopb.bzl (hunk 4 above)
Comment (garethellis0) [reply src/firmware_new/nanopb.bzl:34]: Whoops. I'll look into this tomorrow, but off the top of my head I can't think of any reason any target would depend on both?

Comment (garethellis0) [reply src/firmware_new/nanopb.bzl:34]: Looks like the I just had to re-add `fw_cc_proto`. I think as long as no target depends on both the C and C++ protos at the same time, bazel will make sure there are never any header conflicts via sandboxing. :tada:

**Review by MathewMacDougall — DISMISSED** [review]
Nice work :+1:

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm :fire: :tada: can't wait to use it haha

**Review by jonathanlew — DISMISSED** [review]
LGTM pending CI

**Comment by akhilveeraghanta** [comment]
`firmware_new/boards/frankie_v1/udp_multicast.c` includes the `control_fw.pb.h` need to nuke the `_fw`

**Review by akhilveeraghanta — APPROVED** [review]
:fire:

**Review by jonathanlew — APPROVED** [review]
:100:

**Review by MathewMacDougall — APPROVED** [review]
:shipit:
